### DISCUSSION TITLE
Question about two ways to add built-in actions that need to be executed every block

### DISCUSSION THREAD

**Discussion by riemannulus** [planetarium/libplanet#2301] [Ideas]
DPoS로의 전환 중, 매 블록의 종료 시마다 실행해야 하는 액션이 존재합니다. 이 액션은 각종 상태에 따라 보상을 분배하거나, 대상을 처벌합니다.
이 기능을 구현하는 데 있어 방법이 두 가지가 있습니다.

1. `ActionEvaluator<T>.Execute()`에 직접 구현해 넣는다.
2. `ActionEvaluator<T>.SystemBlockAction` 필드를 만들고 여기에 액션을 등록한 뒤 `ActionEvaluator<T>.Execute()` 에서 호출한다.

1, 2번 모두 기능 호출은 이 곳에서 하려고 합니다.

https://github.com/planetarium/libplanet/blob/2339d1045e712dcf550e686df3a7b32387e9c898/Libplanet/Action/ActionEvaluator.cs#L109-L121

혹시 있다면 1, 2번 안 이외에 다른 의견도 듣고 싶습니다.

**Comment by dahlia** [comment]
Actions are like opcodes for the blockchain state machine.  In other words, they are intended to have some operands, and give users a choice to opt in/out them inside the shared state machine.  However, if a state transformation should be common and deterministic (i.e., users have no choice on operand values or whether opt in or out the transformation) it doesn't have to be an action.  (That's why `IBlockPolicy<T>.BlockAction` makes less sense in fact IMHO.)

To me, it's more natural to implement such state transformation inside `ActionEvaluator<T>` (or it would be much better if you parametrize `ActionEvaluator<T>` so that it can take any kind of *state transformers*, i.e., `Func<IAccountStateDelta, IAccountStateDelta>`).

**Comment by riemannulus** [comment]
해당 안은 https://github.com/planetarium/libplanet/discussions/3119 에서 비슷한 논의를 이어 하고 있습니다.
